### DISCUSSION TITLE
sqlalchemy.exc.ResourceClosedError: This result object is closed.

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#8176] [Usage Questions] [answered]
I am using 1.4.37 async, When I delete many to many relational table data like using synchronization, I receive an error:

these are sync code:
```python
userinfo = session.query(User).filter(User.id == current_user.id)

for i in list(userinfo.first().roles):
    userinfo.first().roles.remove(i)
```
It works

---

these are async code:
```python
user = await session.scalars(
    select(User).where(User.id == current_user.id).options(selectinload(User.roles))
)

for i in list(user.first().roles):
    user.first().roles.remove(i)
```
It doesn't work: `sqlalchemy.exc.ResourceClosedError: This result object is closed.`

**Comment by zzzeek** [comment, marked as answer]
this code is wrong:

```py
for i in list(user.first().roles):
    user.first().roles.remove(i)
```

you may only call `first()` once on a result set.  after that, it's closed.

correct code would resemble:

```py
first_user = user.first()
for i in list(first_user.roles):
    first_user.roles.remove(i)
```

in your previous example, you are using an entirely different API which is the legacy query API, so first() invokes a query each time.

**Reply by wu-clan** [reply]
Great, thank you, zzzeek, I will read more documents
